File: server/services/astrology/rules/grahanDosh.js
/* Grahan Dosha: traditionally read when the Sun or Moon is conjunct Rahu or Ketu
   (an eclipse-like combination in the birth chart). */
'use strict';
const { DISPLAY_HI } = require('../kundaliEngine');
const NODE_HI = { rahu: 'राहु', ketu: 'केतु' };

function evaluate(view) {
  const hits = [];
  for (const lum of ['sun', 'moon']) {
    const L = view.planets[lum];
    for (const node of ['rahu', 'ketu']) {
      const N = view.planets[node];
      if (L.sign === N.sign) {
        const sep = Math.abs(L.degreeInSign - N.degreeInSign);
        hits.push({ luminary: lum, node, separation: +sep.toFixed(1) });
      }
    }
  }
  if (!hits.length) return null;
  const minSep = Math.min(...hits.map((h) => h.separation));
  return {
    detected: true,
    severity: minSep < 5 ? 'high' : minSep < 12 ? 'medium' : 'low',
    confidence: Math.max(0.5, 0.9 - minSep / 25),
    evidence: hits.map((h) =>
      h.luminary.charAt(0).toUpperCase() + h.luminary.slice(1) + ' is in the same sign as ' + h.node +
      ' (about ' + h.separation + '° apart).'),
    evidenceHi: hits.map((h) =>
      DISPLAY_HI[h.luminary] + ' ' + NODE_HI[h.node] + ' के साथ एक ही राशि में है (लगभग ' + h.separation + '° की दूरी पर)।')
  };
}

module.exports = { evaluate };

File: backend-python/app/services/astrology/rules/grahan_dosh.py
"""Grahan Dosha — port of rules/grahanDosh.js: traditionally read when the Sun or
Moon is conjunct Rahu or Ketu (an eclipse-like combination in the birth chart)."""
from ..kundali_engine import DISPLAY_HI

NODE_HI = {'rahu': 'राहु', 'ketu': 'केतु'}


def evaluate(view: dict, now=None, partner=None) -> dict | None:
    hits = []
    for lum in ('sun', 'moon'):
        L = view['planets'][lum]
        for node in ('rahu', 'ketu'):
            N = view['planets'][node]
            if L['sign'] == N['sign']:
                sep = abs(L['degreeInSign'] - N['degreeInSign'])
                hits.append({'luminary': lum, 'node': node, 'separation': round(sep, 1)})
    if not hits:
        return None
    min_sep = min(h['separation'] for h in hits)
    return {
        'detected': True,
        'severity': 'high' if min_sep < 5 else 'medium' if min_sep < 12 else 'low',
        'confidence': max(0.5, 0.9 - min_sep / 25),
        'evidence': [
            h['luminary'].capitalize() + ' is in the same sign as ' + h['node'] + ' (about ' + str(h['separation']) + '° apart).'
            for h in hits
        ],
        'evidenceHi': [
            DISPLAY_HI[h['luminary']] + ' ' + NODE_HI[h['node']] + ' के साथ एक ही राशि में है (लगभग ' + str(h['separation']) + '° की दूरी पर)।'
            for h in hits
        ],
    }
